compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp

```diff
@@ -327,9 +327,21 @@ class LLVMCPUTargetBackend final : public TargetBackend {
                                     .value_or(APInt(64, 0))
                                     .getSExtValue();
 
-      libraryBuilder.addExport(exportOp.getName(), "",
-                               LibraryBuilder::DispatchAttrs{localMemorySize},
-                               llvmFunc);
+      std::string sourceFile = "";
+      if (options_.debugSymbols && !options.dumpedSourcesPath.empty()) {
```

Comment:
Let's not reuse that llvm flag as it means something different (dwarf/pdf debugging symbols) and the two aren't related. We also want this for all backends and don't want the user to need to pass one flag per backend they're using. We really are kind of reaching for a generic executable debug level on the HAL TargetOptions - none/minimal (tracing/annotations)/full (symbols). We could reuse the gcc-style levels 0-3:

```
-g0 | no debug information
-g1 | minimal debug information
-g | default debug information
-g3 | maximal debug information
```

We could start by adding am `--iree-hal-executable-debug-level=N` to TargetOptions and enabling this when >=1 - future changes could then drop the `iree-llvm-debug-symbols` flag and include them at >= 2, etc. I've got an issue for doing this and can do it in a follow-up, but if you're planning to advertise this then we'll want to get it in here as a placeholder even if only used for this.

Reply:
A setting in `TargetOptions` that backends can interpret how they wish SGTM. https://github.com/iree-org/iree/blob/a0943ed99ea31723b839a2234a3070f46d5f2bab/compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h#L46-L51

One thing I'm not sure about is how we want to plumb that information through to `TargetBackend` implementations to make use of. Right now that particular struct is used at a control level above the `TargetBackend` class, such as in here: https://github.com/iree-org/iree/blob/a0943ed99ea31723b839a2234a3070f46d5f2bab/compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp#L71-L72 https://github.com/iree-org/iree/blob/a0943ed99ea31723b839a2234a3070f46d5f2bab/compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp#L105-L108

while the actual methods on classes like `LLVMCPUTarget` only have their own options set: https://github.com/iree-org/iree/blob/a0943ed99ea31723b839a2234a3070f46d5f2bab/compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMCPUTarget.cpp#L221-L223

Reply:
I added a first pass at such a flag. I think `SerializeTargetExecutablesPass` could just carry around `TargetBackend::SerializationOptions` instead of duplicating the lists of options in multiple locations, and some other refactoring could be done. LMK if you want that done ahead of time, as a follow-up, or in this PR.